assimilation_code/modules/io/direct_netcdf_mod.f90

```diff
@@ -856,22 +856,28 @@ subroutine read_variables(ncfile_in, var_block, start_var, end_var, domain)
    allocate(counts(num_dims))
    allocate(slice_start(num_dims))
 
+   slice_start(:) = 1 ! default to read all dimensions start at 1
+
    if (has_unlimited_dim(domain)) then
 
-      slice_start(:) = 1 ! default to read all dimensions start at 1
       counts(num_dims) = 1 ! one slice of unlimited dimesion
```

Comment:
The variable counts may need to be assigned differently for netcdf files with unlimited dimensions versus without. More specifically, the length of the counts array will probably need to be different depending on whether or not there is an unlimited dimension. If there are no unlimited dimensions, the length should be 3. If there is an unlimited dimension (time), then the length should be 4 to include x, y, z, and time.

Reply:
check this out:
I've being trying to break this by making counts and slice_start massive

```fortran
 854    ! number of dimensions and length the variable
 855    num_dims = get_io_num_dims(domain, i)
 856    allocate(counts(num_dims+200))
 857    allocate(slice_start(num_dims+200))
```

Here is the count docs.
https://docs.unidata.ucar.edu/netcdf-fortran/current/f90-variables.html#f90-writing-data-values-nf90_put_var
![Screen Shot 2022-10-06 at 3 54 04 PM](https://user-images.githubusercontent.com/20047007/194425626-6cb14d46-f952-49a4-86b3-cc26a0c27edd.png)

I think if count is too long, but the extra element is 1, that is ok because that is what netcdf is doing anyway:
The netcdf code does stuff like this: (netcdf-fortran-4.5.3/fortran/netcdf_expanded.f90).  

```fortran
1145    function nf90_put_var_1D_EightByteReal(ncid, varid, values, start, count, stride, map)
1146      integer,                         intent( in) :: ncid, varid
1147      real (kind = EightByteReal), dimension(:), &
1148                                       intent( in) :: values
1149      integer, dimension(:), optional, intent( in) :: start, count, stride, map
1150      integer                                      :: nf90_put_var_1D_EightByteReal
1151 
1152      integer, dimension(nf90_max_var_dims) :: localStart, localCount, localStride, localMap
1153      integer                               :: numDims, counter
1154 
1155      ! Set local arguments to default values
1156      numDims                 = size(shape(values))
1157      localStart (:         ) = 1
1158      localCount (:numDims  ) = shape(values)
1159      localCount (numDims+1:) = 1
1160      localStride(:         ) = 1
1161      localMap   (:numDims  ) = (/ 1, (product(localCount(:counter)), counter = 1, numDims - 1) /)
1162 
1163      if(present(start))  localStart (:size(start) )  = start(:)
1164      if(present(count))  localCount (:size(count) )  = count(:)
1165      if(present(stride)) localStride(:size(stride)) = stride(:)
1166      if(present(map))  then
1167        localMap   (:size(map))    = map(:)
1168        nf90_put_var_1D_EightByteReal = &
1169           nf_put_varm_double(ncid, varid, localStart, localCount, localStride, localMap, values)
1170      else if(present(stride)) then
1171        nf90_put_var_1D_EightByteReal = &
1172           nf_put_vars_double(ncid, varid, localStart, localCount, localStride, values)
1173      else
1174        nf90_put_var_1D_EightByteReal = &
1175           nf_put_vara_double(ncid, varid, localStart, localCount, values)
1176      end if
1177    end function nf90_put_var_1D_EightByteReal

```

I'm not totally convinced yet, but figured I would share this update of why it 'works' as is.

Reply:
The only thing I'm a little confused about is the last element of counts. What is the purpose of the extra element, why does it have a default value of 1, and what does that default value of 1 tell NetCDF to do? Similarly, in what cases will values other than 1 be used in the last element, and will this still work if count is too long and the last variable is not 1?

What is a record variable, and why/how does this affect the last element of counts? The documentation says that if the variable is a record variable, then the last element corresponds to the number of records to write, but I'm not really sure what this means. We can talk about all this more at the drop in today too.

Reply:
chatted about this in person today.